doc/design.md

```diff
@@ -0,0 +1,1268 @@
+More code is likely to require templating for the precision (`double` and `single`).
+However, the development cost for this library itself would be rather minor, so there is very little risk involved in supporting it from the beginning.
+
+
+### <a name="design-details-slicing"></a>Slicing
+
+We would like to support slicing as in `numpy` and `xarray`, e.g., `array[2:4,10:20, :]` in the case of slicing `Variable` (or the equivalent with named dimensions in the case of slicing `Dataset`, as supported in `xarray`).
+In both `numpy` and `xarray` the returned object is a view, i.e., data is not extracted.
+
+The ownership and copying discussion in the section [Copy-on-write mechanism](#design-details-copy-on-write-mechanism) is related to such slice-views and implies:
+- Any iterator obtained from a view is invalidated if any non-`const` method on *any other view* to the same `Variable` is called.
+  - The view itself can stay valid since it references the variable, not its data.
+- Contrary to the `numpy` behavior, if the owning `Variable` goes out of scope all views should probably become invalid.
+  To support views that stay valid we would need to wrap all variables in `Dataset` into a `shared_ptr` and the benefit does not seem to justify this additional complexity at this point.
+
+
+## <a name="implementation"></a>Implementation
+
+
+### <a name="implementation-goals-and-non-goals"></a>Goals and non-goals
+
+##### Goals in this implementation phase
+
+1. Provide a simple, lightweight, functional and well-tested library that can replace and enhance most of the workspaces types existing in Mantid.
+1. Reach a point where this library (in combination with some supporting libraries) is useful in practice.
+   We define useful as:
+   - Complete low-level functionality in C++ and Python.
+   - `numpy` interoperability.
+   - Visualization using a slice viewer.
+   - Visualization using a multi-dimensional table viewer.
+   - Instrument visualization using the Mantid instrument viewer.
+   - Can leverage existing Mantid code such as `DataHandling::LoadEventNexus` by a subsequent conversion into a `Dataset`.
+1. Demonstrate usefulness on at least two real examples, ideally bringing them to a production ready stage.
+   Candidates are:
+   - Constant-wavelength (reactor) workflows, which currently use `API::MatrixWorkspace` with length-1 histograms as a workaround.
+   - Imaging workflows, which would benefit from presenting data as a stack of images rather than a list of histograms, i.e., the dimension order would be the opposite of what we are used to in `API::MatrixWorkspace`.
+   - Workflow with parameter scan or polarization analysis to demonstrate new capabilities of extra dimensions and/or multiple data variables.
+1. If applicable, demonstrate performance gain over existing Mantid workflow.
+
+##### Non-goals
+
+Even beyond this implementation phase, we have the following non-goals:
+
+1. It is a non-goal to provide a drop-in replacement for
```

Comment:
I fear there would nevertheless by too many small interface changes, i.e., it would require considerable rollout effort, and would very likely come with some performance penalties.

At least for the initial phases, I am currently considering another strategy: Instead of attempting a "horizontal" rollout we start with a more vertical rollout. For example, after the very basics have been implemented:

- Initially, aim to support 3 production workflows to a 80% or 90% level.
- Implement specific lower level algorithms
- Port workflow-algorithms to use the new algorithms.
- See where the major problems with the design lie, fix them.

After that, we will have a clearer picture. We will be able to tell how many things we actually need to port to support a production workflow. We will have an estimate of how many algorithms are actually needed. We will know if there are any performance benefits. That is, we will be in a position to decide whether continuing with the technique-by-technique rollout, or whether we actually want to port everything using an adapter after all (or whether we want to discard everything and rewrite it in Cobol).

Reply:
I understand what you are suggesting, but I struggle to see how that could ever be released in a way that is acceptable to the 3 facilities that are using this business-critical framework.  We can't realistically have a situation where some algorithms work and others don't with two workspace types that only differ under the hood as far as our users are concerned.

I agree that the adapter or wrapper approach would almost certainly include performance penalties, and we would need to be aware of that when considering the benefits of this design.  There are many aspects I like about this approach and it really excels about adding flexibility.

How about another approach?
- Create a MatrixWorkspaceAdapter
- Identify key core algorithms needed by a slow workflow (e.g. Direct Inelastic)
- Get benchmarks of this process using the old workspaces
- In a branch:
  - Adjust the workspace factory to create all Matrix Workspaces as MatrixWorkspaceAdapter
  - Get wrapper benchmarks
  - Adjust those key algorithms (ConvertUnits, Rebin, Divide, Possibly LoadNexus) to natively use the new workspace
  - Get native new workspace benchmarks

If all is well and good we can then test it more widely, and let the results we find direct us in implementing it wider.

I think we can allow niche algorithms to have a performance hit through working through a wrapper, but we cannot just force deprecation on them by changing the underlying workspace type.

Reply:
As discussed in the [second part of the Non-goals section](https://github.com/mantidproject/dataset/pull/2/files/25a69638c294bc3505802fddccd5bbb8313a1cdf#diff-1ab4257b823a2299454fd7d83cfa11ebR1123), a full rollout (which is what `MatrixWorkspaceAdapter` would amount to) is not doable in my opinion.

My guess is that if you disable compilation of the GUI, 95% of the algorithms, and most of the Python interfaces you may be in a position where you can get a working test case as you describe. Doing that and getting it to work would probably still take something between 3 and 12 months of effort.

That is, it can be done, but regardless of what we learn from the exercise I do not see a solution for how to cover all the disabled parts with the finite amount of available resources.

Furthermore, even if it turns out easier than expected, we gain nearly nothing from it: A new underlying data structure for the existing API defined by `MatrixWorkspace` with all the existing algorithms does not provide any of the benefits of the new data structure. Rather, it combines the drawbacks of the old API with the drawbacks of the new data structure. Again, see the linked *Non-goals* section.

Also note that the `MatrixWorkspaceAdapter` suffers almost the same problem you mention: *"We can't realistically have a situation where some algorithms work and others don't with two workspace types that only differ under the hood as far as our users are concerned."* --- **any** algorithm that makes use of the new features of `Dataset` will "break" that `MatrixWorkspaceAdapter`, i.e., none of the "old" algorithms will work on it anymore.